### PR TITLE
[libc][stdlib] Implement heap sort.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/src/stdlib/qsort_r.cpp [hunk 1]
```diff
@@ -22,8 +22,14 @@ LLVM_LIBC_FUNCTION(void, qsort_r,
   if (array == nullptr || array_size == 0 || elem_size == 0)
     return;
   internal::Comparator c(compare, arg);
-  internal::quicksort(internal::Array(reinterpret_cast<uint8_t *>(array),
-                                      array_size, elem_size, c));
+  auto arr = internal::Array(reinterpret_cast<uint8_t *>(array), array_size,
+                             elem_size, c);
+
+#if LIBC_QSORT_IMPL == LIBC_QSORT_QUICK_SORT
```
Comment (michaelrj-google) [inline libc/src/stdlib/qsort_r.cpp:28]: would it make sense to have this switch be in `qsort_util` and have these functions call something like `internal::sort`?

File: libc/src/stdlib/qsort_r.cpp [hunk 2]
```diff
@@ -22,8 +22,14 @@ LLVM_LIBC_FUNCTION(void, qsort_r,
   if (array == nullptr || array_size == 0 || elem_size == 0)
```
Comment (michaelrj-google) [inline libc/src/stdlib/qsort_r.cpp:22]: this file needs to include `heap_sort.h` if it's going to use it.

File: libc/src/stdlib/heap_sort.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  while (end > 1) {
+    if (start > 0) {
+      --start;
+    } else {
+      --end;
+      array.swap(0, end);
+    }
+
+    size_t root = start;
```
Comment (mysterymath) [inline libc/src/stdlib/heap_sort.h:35]: // Sift start down the heap.

File: libc/src/stdlib/heap_sort.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+LIBC_INLINE void heap_sort(const Array &array) {
+  size_t end = array.size();
+  size_t start = end / 2;
+
+  auto left_child = [](size_t i) -> size_t { return 2 * i + 1; };
+
+  while (end > 1) {
+    if (start > 0) {
+      --start;
```
Comment (mysterymath) [inline libc/src/stdlib/heap_sort.h:29]: // Select the next unheapified element to sift down.

File: libc/src/stdlib/heap_sort.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    } else {
+      --end;
```
Comment (mysterymath) [inline libc/src/stdlib/heap_sort.h:31]: // Extract the max element of the heap, moving a leaf to root to be sifted down.

**Comment by PiJoules** [comment]
Is it possible to to update `qsort_test.cpp` such that it can run both sorting algos over the same test cases?

File: libc/src/stdlib/heap_sort.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      array.swap(0, end);
+    }
+
+    size_t root = start;
+    while (left_child(root) < end) {
+      size_t child = left_child(root);
+      if (child + 1 < end &&
```
Comment (mysterymath) [inline libc/src/stdlib/heap_sort.h:38]: // If there are two children, set child to the greater.

File: libc/src/stdlib/heap_sort.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+          array.elem_compare(child, array.get(child + 1)) < 0)
+        ++child;
+
+      if (array.elem_compare(root, array.get(child)) >= 0)
```
Comment (mysterymath) [inline libc/src/stdlib/heap_sort.h:42]: // If the root is less than the greater child

File: libc/src/stdlib/heap_sort.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        break;
+
+      array.swap(root, child);
```
Comment (mysterymath) [inline libc/src/stdlib/heap_sort.h:45]: // Swap the root with the greater child and continue sifting down.

File: libc/src/stdlib/qsort_r.cpp (hunk 1 above)
Comment (lntue) [reply libc/src/stdlib/qsort_r.cpp]: I refactor `qsort_util.h` and move the selection logic into `qsort_util.h`.

File: libc/src/stdlib/qsort_r.cpp (hunk 2 above)
Comment (lntue) [reply libc/src/stdlib/qsort_r.cpp:22]: Moved the selection logic into `qsort_util.h`.

**Comment by lntue** [comment]
> Is it possible to to update `qsort_test.cpp` such that it can run both sorting algos over the same test cases?

I added the unit tests for `internal::quick_sort` and `internal::heap_sort` directly.

File: libc/src/stdlib/heap_sort.h (hunk 8 above)
Comment (lntue) [reply libc/src/stdlib/heap_sort.h:52]: Done.

File: libc/src/stdlib/heap_sort.h (hunk 7 above)
Comment (lntue) [reply libc/src/stdlib/heap_sort.h:48]: Done.

File: libc/src/stdlib/heap_sort.h (hunk 6 above)
Comment (lntue) [reply libc/src/stdlib/heap_sort.h:43]: Done.

File: libc/src/stdlib/heap_sort.h (hunk 5 above)
Comment (lntue) [reply libc/src/stdlib/heap_sort.h:34]: Done.

File: libc/src/stdlib/heap_sort.h (hunk 4 above)
Comment (lntue) [reply libc/src/stdlib/heap_sort.h:30]: Done.

File: libc/src/stdlib/heap_sort.h (hunk 3 above)
Comment (lntue) [reply libc/src/stdlib/heap_sort.h:39]: Done.

**Review by mysterymath — APPROVED** [review]
LGTM for me, but see if any of the other reviewers have additional comments.

**Review by petrhosek — APPROVED** [review]

**Pull request merged**